File: aoc-2025/day04.py
with open("day04.in") as f:
    inp = [[c == "@" for c in row] for row in f.read().strip().splitlines()]
    N = len(inp)
    assert(len(inp) == len(inp[0]))

def neighbors(x, y):
    for dx in [-1, 0, 1]:
        for dy in [-1, 0, 1]:
            if dx == 0 and dy == 0:
                continue
            i, j = x + dx, y + dy
            if 0 <= i < N and 0 <= j < N:
                yield (i, j)

def movable(inp):
    moved = []
    for x in range(N):
        for y in range(N):
            if not inp[x][y]:
                continue
            count = 0
            for (i, j) in neighbors(x, y):
                if inp[i][j]:
                    count += 1
            if count < 4:
                moved.append((x, y))
    return moved

print(len(movable(inp)))

c = 0
while len(movable(inp)) > 0:
    for (i, j) in movable(inp):
        c += 1
        inp[i][j] = False
print(c)
    

File: aoc-2025/day04.ml
module PointSet = Set.Make (struct
    type t = int * int

    let compare (x, y) (x', y') = if x = x' then compare y y' else compare x x'
  end)

let input =
  open_in "day04.in"
  |> In_channel.input_all
  |> String.trim
  |> String.split_on_char '\n'
  |> List.mapi (fun row line ->
    line
    |> String.to_seq
    |> Seq.mapi (fun col ch -> row, col, ch)
    |> Seq.filter_map (fun (r, c, ch) -> if ch = '@' then Some (r, c) else None)
    |> List.of_seq)
  |> List.concat
  |> PointSet.of_list
;;

let neighbors (x, y) =
  List.map
    (fun (dx, dy) -> x + dx, y + dy)
    [ -1, -1; -1, 0; -1, 1; 0, -1; 0, 1; 1, -1; 1, 0; 1, 1 ]
;;

let is_movable points p =
  let is_paper = Fun.flip PointSet.mem points in
  is_paper p && List.(length (find_all is_paper (neighbors p))) < 4
;;

(* Fix point iteration to remove movable points *)
let rec fix points =
  let movable = PointSet.filter (is_movable points) points in
  if PointSet.is_empty movable then points else fix (PointSet.diff points movable)
;;

let part1 = PointSet.(cardinal (filter (is_movable input) input))
let part2 = PointSet.(cardinal (diff input (fix input)))
let () = Printf.printf "Part 1 = %d, Part 2 = %d\n" part1 part2
